proposals/extensions.md

```diff
@@ -399,54 +399,237 @@ These requirements need more refinement as implementation progresses, and may ne
-The containing static class for an extension declaration is marked with an `[Extension]` attribute.  
+#### Goals
 
-#### Extension types
+The design below allows:
+- roundtripping of extension declaration symbols through metadata (full and reference assemblies),
+- stable references to extension members (xml docs), 
+- local determination of emitted names (helpful for EnC),
+- public API tracking.  
 
-Each extension declaration in source is emitted as an extension declaration in metadata (sometimes referred to as skeleton type).  
-- Its name is unspeakable and determined based on the lexical order in the program.  
-  The name is not guaranteed to remain stable across re-compilation. 
-  Below we use `<>E__` followed by an index. For example: `<>E__2`.  
-- Its type parameters are those declared in source (including attributes).  
-- Its accessibility is public.
+For xml docs, the docID for an extension member is the docID for the extension member in metadata. For example, the docID used in `cref="Extension.extension(object).M(int)"`
+is `M:Extension.<>E__ExtensionGroupingTypeNameForObject.M(System.Int32)` and that docID is stable across re-compilations and re-orderings of extension blocks. Ideally, it would also remain
+stable when constraints on the extension block change, but we didn't find a design that would achieve that without detrimental effect on language design for member conflicts.
+
+For EnC, it is useful to know locally (just by looking at a modified extension member) where the updated extension member is emitted in metadata.  
+
+For public API tracking, more stable names reduce noise. But technically the extension grouping type names should not come into play in such scenarios. When looking at extension member `M`,
+it doesn't matter what is the name of the extension grouping type, what matters is the signature of the extension block it belongs to. The public API signature should not be seen as
+`Extension.<>E__ExtensionGroupingTypeNameForObject.M(System.Int32)` but rather as `Extension.extension(object).M(int)`. In other words, extension members should be seen as having two sets of
+type parameters and two sets of parameters.
+
+#### Overview
+
+Extension blocks are grouped by their CLR-level signature. Each CLR equivalency group is emitted as an **extension grouping type** with a content-based name.
+Extension blocks within a CLR equivalency group are then sub-grouped by C# equivalency. Each C# equivalency group is emitted as an **extension marker type** with a content-based name, nested in its corresponding extension grouping type.
+An extension marker type contains a single **extension marker method** which encodes an extension parameter.
+Each extension member is emitted with a skeleton body in the right extension grouping type, refers back to an extension marker type by its name via an attribute, and is accompanied by a top-level static **implementation method** with a modified signature.    
```

Comment:
>skeleton body [](http://example.com/codeflow?start=40&length=13)

I think we agreed to not use the "skeleton" term. In addition, I do not see any need or benefit in including information about the body in this particular sentence. What body is emitted and where it is emitted is clearly explained later. Therefore, I am suggesting to use the following wording instead: "Declaration of each extension member is emitted in the right extension grouping type, ..."